Essentials/src/main/java/com/earth2me/essentials/ProviderFactory.java

```diff
@@ -0,0 +1,118 @@
+package com.earth2me.essentials;
+
+import io.papermc.lib.PaperLib;
+import net.ess3.provider.Provider;
+import net.essentialsx.providers.ProviderData;
+import net.essentialsx.providers.ProviderTest;
+import org.bukkit.plugin.Plugin;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+
+public class ProviderFactory {
+    private final Map<Class<? extends Provider>, Provider> providers = new HashMap<>();
+    private final Map<Class<? extends Provider>, List<Class<? extends Provider>>> registeredProviders = new HashMap<>();
+    private final Essentials essentials;
+
+    public ProviderFactory(final Essentials essentials) {
+        this.essentials = essentials;
+    }
+
+    public <P extends Provider> P get(final Class<P> providerClass) {
+        if (!this.providers.containsKey(providerClass)) {
+            return null;
+        }
+        //noinspection unchecked
+        return (P) providers.get(providerClass);
+    }
+
+    @SafeVarargs
+    public final void registerProvider(final Class<? extends Provider>... toRegister) {
+        for (final Class<? extends Provider> provider : toRegister) {
+            final Class<?> superclass = provider.getInterfaces().length > 0 ? provider.getInterfaces()[0] : provider.getSuperclass();
+            if (Provider.class.isAssignableFrom(superclass)) {
+                //noinspection unchecked
+                registeredProviders.computeIfAbsent((Class<? extends Provider>) superclass, k -> new ArrayList<>()).add(provider);
+                if (essentials.getSettings().isDebug()) {
+                    essentials.getLogger().info("Registered provider " + provider.getSimpleName() + " for " + superclass.getSimpleName());
+                }
+            }
+        }
+    }
+
+    public void finalizeRegistration() {
+        for (final Map.Entry<Class<? extends Provider>, List<Class<? extends Provider>>> entry : registeredProviders.entrySet()) {
+            final Class<? extends Provider> providerClass = entry.getKey();
+            Class<? extends Provider> highestProvider = null;
+            ProviderData highestProviderData = null;
+            int highestWeight = -1;
+            providerLoop:
+            for (final Class<? extends Provider> provider : entry.getValue()) {
+                try {
+                    final ProviderData providerData = provider.getAnnotation(ProviderData.class);
+                    if (providerData.weight() > highestWeight) {
+                        for (final Method method : provider.getMethods()) {
```

Comment:
suggestion: instead of using a label here, run the loop just to find the method (maybe rename to `test`) and then break out before testing / continuing if not found